### PR TITLE
Django Configuration: Provide Fallback

### LINKED ISSUE #38: Config: Provide default fallback
We should provide an default fallback for the configuration file, and issue an warning (but no error/termination) if the fallback is used.
This would greatly simplify development, as no manual interference with a config file (which will most likely be the example anyway) is needed.

Further details https://github.com/chaosinventory/chaosinventory/issues/38#issuecomment-787454331

### PR DESCRIPTION
**Pull request by margau**
This should fix #38, and hopefully also #32.

We introduce a fallback.cfg file, which is loaded every single time, as suggested in https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.read

We also introduce a warning if the fallback secret is used.

### CODE REVIEW

**Comment by em0lar** [comment]
I would prefer if there is no default secret, even if there is a warning. In my opinion it would be better to have the following possibilities:

* Define the secret directly in the configuration file (no default)
* Define the secret in an extra file. There is a default path, but of course the user can change this)
* If the secret is not defined in the config file and there is no secret file (in the specified path or in the default path) it's auto generated and saved into a file

**Comment by clerie** [comment]
> We introduce a fallback.cfg file, which is loaded every single time, as suggested in https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.read
With this implementation we are in huge trouble. Now we have two locations where we define default. The previous way was using `fallback` in the `get` statements directly. If we prefer using a default file, we should definitely remove these.

**Comment by margau** [comment]
How about letting the fallback-secret as is, warning users when we detect the fallback secret and immediately replacing the fallback secret with a secret generated for this session only? That fits well for dev and CI-purposes (where I'd like to avoid rewriting configuration files), but we warn real users.

**Comment by margau** [comment]
> > We introduce a fallback.cfg file, which is loaded every single time, as suggested in https://docs.python.org/3/library/configparser.html#configparser.ConfigParser.read
> 
> With this implementation we are in huge trouble. Now we have two locations where we define default. The previous way was using `fallback` in the `get` statements directly. If we prefer using a default file, we should definitely remove these.

As stated, we should prefer the fallback statements inside the get-block. But we had problems with  the `configparser.NoSectionError: No section: 'django'`. The sections are initalized using the fallback.cfg, while the real fallback values are provided in the code using get.

**Comment by clerie** [comment]
> and immediately replacing the fallback secret with a secret generated for this session only?
This is a really bad idea I think. The secret is used to salt some stuff in the state and I don't want secrets changing while developing. Just think about sharing your database with others.

**Comment by em0lar** [comment]
> replacing the fallback secret with a secret generated for this session only?

This would be also possible, but then we don't need a default fallback secret, when I understood this correctly. Then we could use a try/except block. 

In general, I think it would be really good to have the possibility to define the secret in an extra file. And if you have this feature, I think it's not far to save the generated secret in this file.

**Comment by clerie** [comment]
> But we had problems with the `configparser.NoSectionError: No section: 'django'`.

Is this problem documented somewhere? I can't find a realated issue.

**Comment by margau** [comment]
> > and immediately replacing the fallback secret with a secret generated for this session only?
> 
> This is a really bad idea I think. The secret is used to salt some stuff in the state and I don't want secrets changing while developing. Just think about sharing your database with others.

https://docs.djangoproject.com/en/3.1/ref/settings/#secret-key says sessions will be invalidated, but not "static" data like the User passwords etc. -  that's something I could live with. Maybe go that way for now, at a later stage we may introduce the persistent secret generation?

**Comment by clerie** [comment]
> it would be really good to have the possibility to define the secret in an extra file.

From the NixOS perspective I would prefer this version too.

**Comment by margau** [comment]
> > But we had problems with the `configparser.NoSectionError: No section: 'django'`.
> 
> Is this problem documented somewhere? I can't find a realated issue.

Not directly. We discussed this in the meeting, and #40 documented the manual way around, as well as #38 was created to propose a better solution. If you start main without a config file in place, you get the NoSectionError.

**Comment by em0lar** [comment]
> Maybe go that way for now, at a later stage we may introduce the persistent secret generation?

I think it's not that complicated to implement this the way I described in https://github.com/chaosinventory/chaosinventory/pull/41#issuecomment-787450964. Is there any reason why we shouldn't do this now the way we want to do it in the long run?

**Comment by margau** [comment]
My suggestion: Leave this as-is, and solve the SECRET-Generation with a persistent file in the future. This is probably a non-trivial topic, as we need to ensure directory availability, file permissions etc.?

**Comment by clerie** [comment]
I think this PR is to big and tries to solve too many different problem. We should divide this a bit as I can't follow.

For me it looks like we have these sub-issues:

- having some default config to build documentation
- having some default config for ci
- fixing this weird `NoSectionError` bug
- getting a convenient way to handle secrets
- allow development without custom config

**Comment by margau** [comment]
> I think this PR is to big and tries to solve too many different problem. We should divide this a bit as I can't follow.
> 
> For me it looks like we have these sub-issues:
> 
> * having some default config to build documentation
> * having some default config for ci
> * fixing this weird `NoSectionError` bug
> * getting a convenient way to handle secrets
> * allow development without custom config

Lets change perspectives:
The only purpose of this PR is allowing the application & documentation to run directly after a git clone, in a minimum viable state. Nothing more, nothing less.

From my perspective, meaningful default values/value generation should not be solved inside this PR, but upon this in the future. This is primary for fixing annoying stuff like #32 in the early project phase.

**Comment by margau** [comment]
Updated the PR:
- we now have only the sections in the fallback.cfg, no values
- SECRET_KEY prepared for future autogeneration/separate key file as proposed by @em0lar

File: src/chaosinventory/settings.py [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import configparser
 import os
+import logging
 from pathlib import Path
 
 config = configparser.RawConfigParser()
+# Read fallback config, initalizing all values needed for basic execution
+config.read_file(open(Path(__file__).parent / "fallback.cfg"))
```
Comment (em0lar) [inline src/chaosinventory/settings.py:8]: With the changes in https://github.com/chaosinventory/chaosinventory/pull/41/commits/98eea83c2edfbec43e75091c5eb38e53a45e6c2b, the fallback configuration file is out of scope of this pull request.

File: src/chaosinventory/fallback.cfg [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[django]
```
Comment (em0lar) [inline src/chaosinventory/fallback.cfg:1]: As said in the other comment, this file should not be added as it's out of scope

File: doc/dev/configuration.rst [hunk 3]
```diff
@@ -0,0 +1,9 @@
+.. _`configuration`:
+
+Configuration
+=============
+
+Chaosinventory uses Pythons `ConfigParser <https://docs.python.org/3/library/configparser.html>`_ for user configuration. The used configuration file may be loaded from different directories.
+
+Chaosinventory always loads the :code:`src/chaosinventory/fallback.cfg` first, which initalizes the empty sections. Every use of configuration data must work with only the fallback data available.
```
Comment (em0lar) [inline doc/dev/configuration.rst:8]: As said in the other comment, this file should not be added as it's out of scope and not needed

**Review by em0lar — CHANGES_REQUESTED** [review]

**Review by em0lar — APPROVED** [review]

**Review by clerie — CHANGES_REQUESTED** [review]
Ok, this looks slightly better now. But I think these are still two separate PRs. One is the documentation change and one is the code change. A PR labeled with `documentation` shouldn't contain any changes on the actual code base and definitely nothing at some security critical sections like secrets.

**Review by clerie — APPROVED** [review]

**Review by e1mo — APPROVED** [review]

**Pull request merged**